lib/actionUtils.ts

```diff
@@ -0,0 +1,112 @@
+import { OpenAPIV3_1 } from "openapi-types";
+
+export function isChoiceRequired(
+  schema: OpenAPIV3_1.SchemaObject,
+  parentRequired: boolean = true
+): boolean {
+  /** Determines whether there's a choice to be made by the AI for
+   * a parameter. There's no choice if it is required and has 1 enum
+   * value (or all nested children are like this) **/
+  if (!parentRequired) return true;
+  if (schema.type !== "object") {
+    return !(schema.enum && schema.enum.length === 1) ?? true;
+  }
+
+  const requiredKeys = schema.required || [];
+
+  for (let key in schema.properties) {
+    const isKeyRequired = requiredKeys.includes(key);
+    const childSchema = schema.properties[key] as OpenAPIV3_1.SchemaObject;
+
+    // If any nested object requires choice, then choice is required.
+    if (isChoiceRequired(childSchema, isKeyRequired)) {
+      return true;
+    }
+  }
+
+  // If no choice required in any properties, then no choice required.
+  return false;
+}
+
+export function getFilledNoChoiceRequiredFields(
+  schema: OpenAPIV3_1.SchemaObject | undefined,
+  parentRequired: boolean = true
+): any {
+  /** Gets an object with filled fields where no choice is required by the AI.
+   *
+   * Used after the AI call, before API call, to get the parameters
+   * where no choice is required, but the parameters need to be filled
+   * for the API call to succeed. **/
+  // TODO: Doesn't deal with arrays yet, only deals with nested objects
+  if (!schema || schema.type !== "object" || !parentRequired) {
+    return null;
+  }
+
+  let filledObject: any = {};
+  const requiredKeys = schema.required || [];
+
+  for (let key in schema.properties) {
+    const isKeyRequired = requiredKeys.includes(key);
+    const childSchema = schema.properties[key] as OpenAPIV3_1.SchemaObject;
+
+    // If key is required and has only one enum value, add it into the filled object.
+    if (isKeyRequired && childSchema.enum && childSchema.enum.length === 1) {
+      filledObject[key] = childSchema.enum[0];
+    } else if (childSchema.type === "object") {
+      // Else if the property is an object, then recursively fill its parameters.
+      const filledChildObject = getFilledNoChoiceRequiredFields(
+        childSchema,
+        isKeyRequired
+      );
+      if (filledChildObject && Object.keys(filledChildObject).length > 0) {
+        filledObject[key] = filledChildObject;
+      }
+    }
+  }
+
+  return filledObject;
+}
+
+export function fillNoChoiceRequiredParams<
+  Params extends Record<string, unknown>,
+  Output extends Params
+>(paramsFromAI: Params, schema: OpenAPIV3_1.SchemaObject): Output {
```

Comment:
I think this either needs a comment or a new name e.g. `combineAIandAutomaticParams`, that's rubbish but you get my drift.
 it took me a while to figure out that you pull out "fillOutable" params from the schema then combine them with the params from the AI